### PR TITLE
[luci] Revise shape inference for CircleSlice

### PR DESCRIPTION
**Pull request by seanshpark**
This will revise shape inference for CircleSlice Op to skip non-constant inputs.

ONE-DCO-1.0-Signed-off-by: SaeHie Park <saehie.park@gmail.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
For #8273, from draft #8219

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp [hunk 1]
```diff
@@ -1312,8 +1312,13 @@ loco::NodeShape infer_slice(const luci::CircleSlice *node)
...
   auto input_shape = luci::shape_get(node->input()).as<loco::TensorShape>();
 
-  auto const_begin = loco::must_cast<luci::CircleConst *>(node->begin());
-  auto const_size = loco::must_cast<luci::CircleConst *>(node->size());
+  auto const_begin = dynamic_cast<luci::CircleConst *>(node->begin());
+  auto const_size = dynamic_cast<luci::CircleConst *>(node->size());
+  if (const_begin == nullptr || const_size == nullptr)
+  {
+    // NOTE shape depends on values of begin and size
+    return use_own(node);
```
Comment (llFreetimell) [inline compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: This case seems to be dynamic shape.

If we don't know `begin()` or `size()`, we cannot decide each dimension values.
However, there is one thing we can decide, which is rank :D

For example, let's assume following situation.
- `input()`, `begin()`, `size()` are not `CircleConst`
- shape of  `input()` : [1, 2, 3, 4]
- shape of  `begin()` : [4]
- shape of `size()` : [4]
Then, shape of `CircleSlice` would be [?, ?, ?, ?].
This can be represented as following NodeShape.
```cpp
  loco::TensorShape output_shape;
  output_shape.rank(4);
  output_shape.dim(0).unset();
  output_shape.dim(1).unset();
  output_shape.dim(2).unset();
  output_shape.dim(3).unset();
```

In conclusion, current PR can be updated as following.
```suggestion
    loco::TensorShape output_shape;
    output_shape.rank(input_shape.rank());
    for (uint32_t i = 0;i < output_shape.rank(); ++i)
      output_shape.dim(i).unset();
    return output_shape;
```

**Comment by llFreetimell** [comment]
Would you add testcase for dynamic shape inference in `luci/service/src/Nodes/CircleSlice.test.cpp`?
Following is one of the example!

```cpp
#include "luci/Service/CircleShapeInference.h"

TEST(ShapeInferenceTest, Slice_dynamic)
{
  luci::sinf::Algorithm shape_inference_algo;

  auto g = loco::make_graph();
  auto slice_input = g->nodes()->create<luci::CircleRelu>();
  auto slice_begin = g->nodes()->create<luci::CircleRelu>();
  auto slice_size = g->nodes()->create<luci::CircleRelu>();
  auto slice = g->nodes()->create<luci::CircleSlice>();

  slice_input->shape({1,2,3,4});
  slice_input->shape_status(luci::ShapeStatus::VALID);
  slice_begin->shape({4});
  slice_begin->shape_status(luci::ShapeStatus::VALID);
  slice_size->shape({4});
  slice_size->shape_status(luci::ShapeStatus::VALID);

  slice->input(slice_input);
  slice->begin(slice_begin);
  slice->size(slice_size);

  auto output_shape = slice->accept(&shape_inference_algo);

  ASSERT_FALSE(output_shape.dim(0).known());
  ASSERT_FALSE(output_shape.dim(1).known());
  ASSERT_FALSE(output_shape.dim(2).known());
  ASSERT_FALSE(output_shape.dim(3).known());
}
```

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp (hunk 1 above)
Comment (llFreetimell) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: FYI, this is why `Slice` in [initial model](https://github.com/Samsung/ONE/issues/8217#issuecomment-1011694606) has [1, 1, 1] :D
If we dump this with `tfldump`, `shape_signature` may have `[-1, -1, -1]`!

Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]:
```
    for (uint32_t i = 0;i < output_shape.rank(); ++i)
      output_shape.dim(i).unset();
    return output_shape;
```
This will remove current shape info read from the file.
I don't think we have any other nodes do this. Or is there?
Using ` return use_own(node);` was to preserve what was from the file.
Can you explain more about why unset dim values?

Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: > If we dump this with tfldump, shape_signature may have [-1, -1, -1]!

I'll check this

Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: Your are right!
```
T(2:35) FLOAT32 (1, 1, 1) (-1, -1, -1) B(67) LSTM_2df696de/rnn/while/TensorArrayV2Write/TensorListSetItem6
T(2:38) FLOAT32 (1, 1, 1) (-1, -1, -1) B(70) LSTM_2df696de/rnn/while/TensorArrayV2Write/TensorListSetItem9
```

Comment (llFreetimell) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: > I don't think we have any other nodes do this. Or is there?

I had a plan with #5683 but nothing done yet because it was not that urgent and no real use case...

> Can you explain more about why unset dim values?

`unset()` makes the dimension `unknown`.
i.e., `output_shape.dim(0).unset()` means `[?]`
https://github.com/Samsung/ONE/blob/ca224b4a329382bf146c06b2c312881f238cdfd4/compiler/loco/include/loco/IR/Dimension.h#L59-L63

Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: There's a problem: `Optimized graph is invalid` is dropped.
output node shape vs graph output shape differes: `[1,?,?] vs [5,3,20]`

**Comment by seanshpark** [comment]
> There's a problem: Optimized graph is invalid is dropped.
output node shape vs graph output shape differes: [1,?,?] vs [5,3,20]

@llFreetimell , can you guess what may be the problem of this?
(1) As we don't eval the network, we can't get calcuate correct shape
(2) some nodes needs fix
(3) ?

There's a comment in `own_shape`
```cxx
loco::TensorShape own_shape(const luci::CircleNode *node)
{
  loco::TensorShape shape;
  shape.rank(node->rank());
  for (uint32_t r = 0; r < node->rank(); ++r)
  {
    // Shape inference rules in this file did not consider unknown dimension.
    // If some node has unknown dimension, 0 is inserted and wrong shape
    // inference was done as a result.
    // To fix this, new shape inference algorithm is being implemented.
    // Until new inference algorithm is fully implemented, unknown dimension
    // would be represented as 1 along with TFLite expression.
    shape.dim(r) = node->dim(r).known() ? node->dim(r).value() : 1;
  }
  return shape;
}
```

is this comment still valid? that is do we need more changes?

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp (hunk 1 above)
Comment (llFreetimell) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:1320]: Looking the graph but I cannot understand why `concat( [?,?,?], [?,?,?], [1,3,20] ) = [5, 3, 20]`.... -_-;;

I will look more the model with https://github.com/Samsung/ONE/issues/8217#issuecomment-1011860364...

**Comment by llFreetimell** [comment]
> can you guess what may be the problem of this?

First we should investigate why [5, 3, 20] is inferred...
- If the original network is something wrong, we should update the incorrect shape to our inferred shape.
  - https://github.com/Samsung/ONE/blob/ca224b4a329382bf146c06b2c312881f238cdfd4/compiler/luci/pass/src/CircleShapeInferencePass.test.cpp#L187
  - When operation has wrong shape, it is updated by our shape. But if graph output has wrong shape, nothing done for now.
- If the original network is correct using other hints, we may add the hints.

> is this comment still valid? that is do we need more changes?

This is still valid..!

- Importing dynamic shape : Done
- Handling dynamic shape : Not done
  - Let's assume there is a shape [1, 2, ?, 4]
  - Before importing dynamic shape : [1, 2, 1, 4]
  - After importing dynamic shape : [1, 2, ?, 4] 
  - If we do not add `shape.dim(r) = node->dim(r).known() ? node->dim(r).value() : 1;`, `dim(3).value()` returns 0, not 1.
  - It breaks some of shape inference logic

**Comment by seanshpark** [comment]
> Handling dynamic shape : Not done 

OK got it. Recurrent may needs this.

**Comment by llFreetimell** [comment]
I am looking the model, but before that, there is one way to get a model quickly.
Using `--disable_validation`....! 
(You may know already but just for information... T.T)

**Comment by seanshpark** [comment]
> Using --disable_validation....!

Yup :) I wasn't in a hurry for the model, well anyway with this option got this; which is also strange.

image

**Comment by seanshpark** [comment]
I think it would be better to continue with new shape inference. closing :)
